src/Access/GSSAcceptor.cpp

```diff
@@ -0,0 +1,469 @@
+        }
+
+        minor_status = 0;
+        major_status = gss_acquire_cred(
+            &minor_status,
+            acceptor_name,
+            GSS_C_INDEFINITE,
+            GSS_C_NO_OID_SET,
+            GSS_C_ACCEPT,
+            &acceptor_credentials_handle,
+            nullptr,
+            nullptr
+        );
+
+        if (GSS_ERROR(major_status))
+        {
+            const auto messages = extractStatusMessages(major_status, minor_status, GSS_C_NO_OID);
+            throw Exception("gss_acquire_cred() failed" + (messages.empty() ? "" : ": " + messages), ErrorCodes::KERBEROS_ERROR);
+        }
+    }
+}
+
+String GSSAcceptorContext::processToken(const String & input_token, Poco::Logger * log)
+{
+    std::scoped_lock lock(gss_global_mutex);
+
+    String output_token;
+
+    try
+    {
+        if (is_ready || is_failed || context_handle == GSS_C_NO_CONTEXT)
+            reset();
+
+        gss_buffer_desc input_token_buf;
+        input_token_buf.length = input_token.size();
+        input_token_buf.value = const_cast<char *>(input_token.c_str());
+
+        gss_buffer_desc output_token_buf;
+        output_token_buf.length = 0;
+        output_token_buf.value = nullptr;
+
+        gss_name_t initiator_name = GSS_C_NO_NAME;
+        gss_OID mech_type = GSS_C_NO_OID;
+        OM_uint32 flags = 0;
+
+        SCOPE_EXIT({
+            if (initiator_name != GSS_C_NO_NAME)
+            {
+                OM_uint32 minor_status = 0;
+                [[maybe_unused]] OM_uint32 major_status = gss_release_name(
+                    &minor_status,
+                    &initiator_name
+                );
+                initiator_name = GSS_C_NO_NAME;
+            }
+
+            OM_uint32 minor_status = 0;
+            [[maybe_unused]] OM_uint32 major_status = gss_release_buffer(
+                &minor_status,
+                &output_token_buf
+            );
+        });
+
+        OM_uint32 minor_status = 0;
+        OM_uint32 major_status = gss_accept_sec_context(
+            &minor_status,
+            &context_handle,
+            acceptor_credentials_handle,
+            &input_token_buf,
+            GSS_C_NO_CHANNEL_BINDINGS,
+            &initiator_name,
+            &mech_type,
+            &output_token_buf,
+            &flags,
+            nullptr,
+            nullptr
+        );
+
+        if (major_status == GSS_S_COMPLETE)
+        {
+            if (!params.mechanism.empty() && !equalMechanisms(params.mechanism, mech_type))
+                throw Exception("gss_accept_sec_context() succeeded, but: the authentication mechanism is not what was expected", ErrorCodes::KERBEROS_ERROR);
+
+            if (flags & GSS_C_ANON_FLAG)
+                throw Exception("gss_accept_sec_context() succeeded, but: the initiator does not wish to be authenticated", ErrorCodes::KERBEROS_ERROR);
+
+            std::tie(user_name, realm) = extractNameAndRealm(initiator_name);
```

Comment:
I believe the credentials are not the right place for all that stuff.
The function `ExternalAuthenticators::checkKerberosCredentials()` currently only checks realm, I think it can do the rest of the logic too. I mean we could add one more overload for `Context::setUser()`, which will get credentials instead of `user_name`+`password`, like the following:

```
void Context::setCredentials(const Credentials & credentials)
{
   // main function, does authentication and get user's name (either from credentials or from the authentication process
}

void Context::setUser(const String & user_name, const String & password)
{
  /// wrap user_name and password into Credentials and then call setCredentials
}
```
}